### PR TITLE
Adding details and making some things more explicit in docs/demo.ipynb

### PR DESCRIPTION
**Pull request by cdagnino**
I read through `demo.ipynb` and tried to clarify the points in which I was having doubts. 

I have two doubts to finish this PR:

1. I'm still unsure of what the np.array([[1]]) does in the `sample` function. If I got it correctly, it's selecting a particular entry in the `known_alphas` dictionary. Is it the context? Since there's no context here, Is the `1` just a default value then? 
2. In the update from the reward stream, why is it using `-2` in `agent.update(np.atleast_1d(reward_stream.getvalue()[-2]))` and not the last element `0` ?

@rishi-kulkarni  I hope it helps!

### CODE REVIEW

**Comment by rishi-kulkarni** [comment]
Thanks for this contribution. Few notes:

1. Under the hood, a non-contextual bandit is supported by ordinary learners that have only one feature, a global intercept. The 1.0 represents that intercept.
2. I actually have been thinking this example is fairly overcomplicated for a demo, primarily due to the stdout capturing. Could you look at having `EmailOracle` append to a list and use that list to update the bandit? That would remove some of the "window dressing" and focus the notebook on the mechanics of the bandit.

**Comment by cdagnino** [comment]
1. Got it. Updated the explanation. For non contextual bandits it might be cleaner to leave it empty (sample without any context) and then capture that `None` and replace it with [[1]].
2. I took out the standard output dependency and put it inside a `rewards_history` list within the oracle class

**Comment by rishi-kulkarni** [comment]
Yeah, I actually had it that way originally but it made static typing a bit of a nightmare. The current implementation has Agents and ContextualAgents, but under the hood they use the same learners, policies, and Arm class. The Agent fills in the global intercept for you, but the Arm itself doesn't know it's part of a non-contextual bandit. Generally, users shouldn't be calling Arm methods directly unless they know what they're doing, though, so I'm okay with it as it is.

Thanks for these changes. I'll give it a closer review this evening.

File: docs/notebooks/demo.ipynb [hunk 1]
```diff
@@ -77,7 +100,10 @@
    "metadata": {},
    "source": [
     "\n",
-    "Additionally, we need to define functions that represent the action space and function(s) to compute reward. Because we consider an output of `T` a success and an output of `F` a failure, we'll have our reward function return the value of the `T` column of our `clf.classes_` array. "
+    "Additionally, we need to define functions that represent the action space and function(s) to compute the reward. Because we consider an output of `C` a success and an output of `N` a failure, we'll have our reward function return the value of the `C` column of our `clf.classes_` array. The `.classes_` won't be initialized until the learner learns from the data at least once, but you can see the order of the classes in `clf.alphas.keys()`\n",
+    "\n",
+    "*Aside: [as defined in the `Arm` class](https://github.com/bayesianbandits/bayesianbandits/blob/8378f2566153285a224f42ca9667e2d22f9f472b/bayesianbandits/_arm.py#L108), a reward function takes samples from the posterior as input: `reward_function(self.learner.sample(X, size))`*\n",
```
Comment (rishi-kulkarni) [inline docs/notebooks/demo.ipynb:105]: Might be worth linking the type signature here: https://github.com/bayesianbandits/bayesianbandits/blob/8378f2566153285a224f42ca9667e2d22f9f472b/bayesianbandits/_arm.py#L15

A reward function takes samples from the posterior as input and returns either a float or an array of floats.

File: docs/notebooks/demo.ipynb [hunk 2]
```diff
@@ -98,18 +126,18 @@
...
   {
    "attachments": {},
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Now we can define our learning agent. To do so, we will need the main decorator from the `bayesianbandits` library - `@bandit`. The `@bandit` decorator synthesizes the bandit API on our experiment class.\n",
+    "Now we can define our learning agent. We first define the arms. Each arm requires an `action_token` (the token that is returned when that particular arm is pulled), a reward function and a learner. This `learner` defines how each arm will learn how rewarding its action is, \n",
     "\n",
-    "The `@bandit` takes two arguments, the `learner` which defines how each arm will learn how rewarding its action is, and `policy`, which determines how the agent chooses which arm to pull. In this case, we'll use `thompson_sampling`, which takes no parameters."
+    "With those arms, we can then define an `Agent`. An `Agent` is defined by a list of arms and a policy, which determines how the agent chooses which arm to pull. In this case, we'll use `thompson_sampling`, which takes no parameters."
```
Comment (rishi-kulkarni) [inline docs/notebooks/demo.ipynb:140]: We're using the `ThompsonSampling` class now, correct? Not the old `thompson_sampling` closure function?

File: docs/notebooks/demo.ipynb [hunk 3]
```diff
@@ -191,6 +224,8 @@
    "source": [
     "import matplotlib.pyplot as plt\n",
     "\n",
+    "print(\"Known alphas dictionary: \", agent.arms[1].learner.known_alphas_)\n",
```
Comment (rishi-kulkarni) [inline docs/notebooks/demo.ipynb:227]: Not sure this is adding much clarity, what do you think?

**Review by rishi-kulkarni — COMMENTED** [review]
This is looking quite good. Few comments for your consideration:

File: docs/notebooks/demo.ipynb (hunk 1 above)
Comment (cdagnino) [reply docs/notebooks/demo.ipynb:105]: Done!

File: docs/notebooks/demo.ipynb (hunk 2 above)
Comment (cdagnino) [reply docs/notebooks/demo.ipynb:140]: Corrected

File: docs/notebooks/demo.ipynb (hunk 3 above)
Comment (cdagnino) [reply docs/notebooks/demo.ipynb:227]: Agreed, I took out the `known_alphas` reference in the "aside" part of the text too

**Review by rishi-kulkarni — APPROVED** [review]
Thanks, this looks quite good!

**Pull request merged**